chore(dataobj): apply predicates sequentially

ashwanthgoli opened the pull request:
**What this PR does / why we need it**:

Updates the dataset reader to accept multiple predicates that get evaluated sequentially. When evaluating a predicate in the list, only columns that are required to evaluate it are loaded. Rows are filtered at each step of the evaluation with subsequent predicate evaluation having to fill the reduced row range.

For now the reader expects the caller to order predicates in a way to benefit from selectivity ~~and to ensure that a column is not referenced in multiple predicates from the list <- this might change in a follow-up commit or pr~~


```
goos: darwin
goarch: arm64
pkg: github.com/grafana/loki/v3/pkg/logql/bench
cpu: Apple M1 Pro
                                                                                                         │ /tmp/main.txt │            /tmp/seq.txt            │
                                                                                                         │    sec/op     │   sec/op     vs base               │
LogQL/query={service_name="grafana",_env="prod",_region="us-west-2"}_[BACKWARD]/kind=log/store=dataobj-8     545.4m ± 7%   498.3m ± 5%  -8.64% (p=0.000 n=30)

                                                                                                         │   /tmp/main.txt    │                /tmp/seq.txt                │
                                                                                                         │ kilobytesProcessed │ kilobytesProcessed  vs base                │
LogQL/query={service_name="grafana",_env="prod",_region="us-west-2"}_[BACKWARD]/kind=log/store=dataobj-8         102.78k ± 0%          54.28k ± 0%  -47.19% (p=0.000 n=30)
```


**Which issue(s) this PR fixes**:
Fixes #<issue number>

**Special notes for your reviewer**:

**Checklist**
- [ ] Reviewed the [`CONTRIBUTING.md`](https://github.com/grafana/loki/blob/main/CONTRIBUTING.md) guide (**required**)
- [ ] Documentation added
- [ ] Tests updated
- [ ] Title matches the required conventional commits format, see [here](https://www.conventionalcommits.org/en/v1.0.0/)
  - **Note** that Promtail is considered to be feature complete, and future development for logs collection will be in [Grafana Alloy](https://github.com/grafana/alloy). As such, `feat` PRs are unlikely to be accepted unless a case can be made for the feature actually being a bug fix to existing behavior.
- [ ] Changes that require user attention or interaction to upgrade are documented in `docs/sources/setup/upgrade/_index.md`
- [ ] If the change is deprecating or removing a configuration option, update the `deprecated-config.yaml` and `deleted-config.yaml` files respectively in the `tools/deprecated-config-checker` directory. [Example PR](https://github.com/grafana/loki/pull/10840/commits/0d4416a4b03739583349934b96f272fb4f685d15)

owen-d reviewed (commented):
I'd like to get @rfratto's opinion on this, but I like it 👍 . Let's add some benchmark cases (if they don't exist) which test querying a list of predicates where each successive predicate removes a more of the remaining rows (e.g. p0 removes 50% of rows and p1 removes half of the remaining rows, etc)


It's worth noting predicate filtering happens in two phases:
1) via page removal (`buildPredicateRanges`) in `initDownloader`. This removes whole pages based on page metadata (e.g. min/max bounds)
2) via row removal, applied while loading rows directly were not filtered out by (1)

It'll be nice to have benchmarks for both, but (1) is net-new whereas (2) should already exist (page filtering is not a new optimization).

rfratto commented on pkg/dataobj/internal/dataset/reader.go:
```diff
@@ -125,38 +126,79 @@ func (r *Reader) Read(ctx context.Context, s []Row) (n int, err error) {
-			size := s[i].SizeOfColumns(r.primaryColumnIndexes)
+		// sequentially apply rest of the predicates.
+		for i, p := range r.opts.Predicates {
+			columns, idxs, err := r.predicateColumns(p)
```
Calling this for every predicate per read call is fine for now, but let's keep an eye out for whether we need to cache these results depending on how many read calls we end up having 👍

rfratto reviewed (commented):
LGTM, I don't have any blocking comments :)

ashwanthgoli commented on pkg/dataobj/internal/dataset/reader.go:
```diff
@@ -125,38 +126,79 @@ func (r *Reader) Read(ctx context.Context, s []Row) (n int, err error) {
+
+			var count int
+
+			// read the requested number of rows for the first predicate.
+			if i == 0 {
+				count, err = r.inner.ReadColumns(ctx, columns, s[:readSize])
+				if err != nil && !errors.Is(err, io.EOF) {
+					return n, err
+				} else if count == 0 && errors.Is(err, io.EOF) {
+					return 0, io.EOF
+				}
+
+				rowsRead = count
+			} else {
+				count, err = r.inner.Fill(ctx, columns, s[:readSize])
+				if err != nil && !errors.Is(err, io.EOF) {
+					return n, err
+				} else if count != readSize {
+					return n, fmt.Errorf("failed to fill rows: expected %d, got %d", len(s), count)
+				}
+			}
```
There might be benefits to the downloader being aware of the current set of columns being used to evaluate a predicate. Because we'd want to first read through all the pages (for the row range) of the current predicate columns rather than doing a round-robin iter over all primary column pages.

If any of the later predicate columns is large (log column), even a few pages loaded from it might occupy a large chunk of targetCacheSize, but it may never get used if filtered out or it could get replaced when loading the next set of pages for evaluating the current predicate.

ashwanthgoli commented on pkg/dataobj/internal/dataset/reader.go:
```diff
@@ -125,38 +126,79 @@ func (r *Reader) Read(ctx context.Context, s []Row) (n int, err error) {
-			size := s[i].SizeOfColumns(r.primaryColumnIndexes)
+		// sequentially apply rest of the predicates.
+		for i, p := range r.opts.Predicates {
+			columns, idxs, err := r.predicateColumns(p)
```
tackled in https://github.com/grafana/loki/pull/17283/commits/5b5805201a2aaa6f033f2d4b8fd3aa4ced764816

ashwanthgoli commented on pkg/dataobj/internal/dataset/reader.go:
```diff
@@ -125,38 +126,79 @@ func (r *Reader) Read(ctx context.Context, s []Row) (n int, err error) {
-	if r.opts.Predicate == nil {
-		// If there's no predicate, all rows are valid.
+	// If there are no predicates, read all columns in the dataset
+	if len(r.opts.Predicates) == 0 {
```
extracted in https://github.com/grafana/loki/pull/17283/commits/4b69cc84cddb90653e65ff6fa3c9cd667b98f555

ashwanthgoli commented:
updated the pr to add benchmarks for the reader. I have added one to measure gains from sequential predicate evaluation alone (not using page pruning)
benchmark columns here are:
- combined predicate
- sequential predicates with low selectivity first
- sequential predicates with high selectivity first

~5% reduction in cpu where both predicates are int64 columns. i think the gains depend a lot depends on data distribution, we could see better number from moving larger columns (log message for ex) to the right.

```
goos: darwin
goarch: arm64
pkg: github.com/grafana/loki/v3/pkg/dataobj/internal/dataset
cpu: Apple M1 Pro
                     │  combined   │                low                 │                high                │
                     │   sec/op    │   sec/op     vs base               │   sec/op     vs base               │
PredicateExecution-8   97.65m ± 2%   98.32m ± 3%  +0.68% (p=0.043 n=10)   92.62m ± 1%  -5.16% (p=0.000 n=10)

                     │   combined   │                 low                  │                 high                 │
                     │     B/op     │     B/op      vs base                │     B/op      vs base                │
PredicateExecution-8   330.0Ki ± 0%   269.9Ki ± 0%  -18.21% (p=0.000 n=10)   241.9Ki ± 0%  -26.69% (p=0.000 n=10)

                     │  combined   │                 low                 │                high                │
                     │  allocs/op  │  allocs/op   vs base                │  allocs/op   vs base               │
PredicateExecution-8   11.82k ± 0%   13.82k ± 0%  +16.94% (p=0.000 n=10)   10.98k ± 0%  -7.10% (p=0.000 n=10)
```

owen-d reviewed (approved):
Thanks, LGTM!
